### DISCUSSION TITLE
Only 2 Piranha modules are listed

### DISCUSSION THREAD

**Discussion by sormuras** [sormuras/modules#18] [Q&A]
https://twitter.com/Thi_Hup/status/1371049989758197761 by @Thihup

> What I don't get is why there is only 2 Piranha modules even though we publish over 50 modules
> And they got released all at once if I remember correctly

Let's see, why there are only two modules listed. The two modules are (as of today):

```properties
cloud.piranha.micro=https://repo.maven.apache.org/.../piranha-micro-21.3.0.jar
cloud.piranha.micro.loader=https://repo.maven.apache.org/.../piranha-micro-loader-21.3.0.jar
```

Links
- GitHub _"Add module-info.java to all our JARs"_ https://github.com/piranhacloud/piranha/pull/1040/files
- Maven Central https://repo.maven.apache.org/maven2/cloud/piranha/

**Comment by Thihup** [comment]
Exactly. Is there something I can do to help the investigation?

**Comment by sormuras** [comment]
A (local) trace of Maven GA `cloud.piranha:piranha-api` looks like this:

```
Scanning 74227 files in G:\zip\s3\java9plusadoption...
Scan[G=cloud.piranha, G2=cloud.piranha, A=piranha-api, GA=cloud.piranha:piranha-api, V=19.12.0, module=-, kind=?]
Scan[G=cloud.piranha, G2=cloud.piranha, A=piranha-api, GA=cloud.piranha:piranha-api, V=20.1.2, module=-, kind=?]
Scan[G=cloud.piranha, G2=cloud.piranha, A=piranha-api, GA=cloud.piranha:piranha-api, V=20.2.0, module=-, kind=?]
Scan[G=cloud.piranha, G2=cloud.piranha, A=piranha-api, GA=cloud.piranha:piranha-api, V=20.3.0, module=-, kind=?]
Scan[G=cloud.piranha, G2=cloud.piranha, A=piranha-api, GA=cloud.piranha:piranha-api, V=20.4.0, module=-, kind=?]
Scan[G=cloud.piranha, G2=cloud.piranha, A=piranha-api, GA=cloud.piranha:piranha-api, V=20.5.0, module=-, kind=?]
Scan[G=cloud.piranha, G2=cloud.piranha, A=piranha-api, GA=cloud.piranha:piranha-api, V=20.6.0, module=-, kind=?]
Scan[G=cloud.piranha, G2=cloud.piranha, A=piranha-api, GA=cloud.piranha:piranha-api, V=20.7.0, module=-, kind=?]
Scan[G=cloud.piranha, G2=cloud.piranha, A=piranha-api, GA=cloud.piranha:piranha-api, V=20.8.0, module=-, kind=?]
  5.396.283 lines
  2.106.807 distinct scan lines
    209.366 artifacts
    194.096 JAR files are plain
     10.446 JAR files with Automatic-Module-Name
      4.824 JAR files with module-info.class
      3.893 distinct modules
      3.000 unique modules
```

Comparing with contents (and metadata) of https://repo.maven.apache.org/maven2/cloud/piranha/piranha-api/ ... reveals:
- missing versions (not a single entry starting for 2021 ... there should be at least 3 of them)
- missing module names (like `...V=20.8.0, module=-, kind...`)

**Reply by sormuras** [reply]
🤔
Possible reasons for the missing assets and names include:
- https://github.com/sandermak/modulescanner didn't run at the time of  your upload (on Sonatype's hardware)
- modulescanner could not scan the uploaded modular JAR file (Java class file version issues?)
- Sonatype didn't publish the related CSV file(s) to the shared S3 bucket
- this project's Scanner implementation is faulty

**Reply by sormuras** [reply]
`jar --describe-module --file piranha-api-21.3.0.jar` yields

### JDK 16
```
cloud.piranha.api@21.3.0 jar:file:///.../piranha-api-21.3.0.jar!/module-info.class
exports cloud.piranha.api
requires java.base mandated
```

### JDK 11
```
java.lang.module.InvalidModuleDescriptorException: Unsupported major.minor version 59.0
```

This is it. At least this module descriptor is compiled to target Java release 15. I guess, the other "over 50 modules" are also targeting Java 15. The `modulescanner` program that analyzes publication at Maven Central runs on Java 11. See https://github.com/sandermak/modulescanner/issues/15 for details.

**Reply by sormuras** [reply]
Is it possible to add `--release 9` (or `--release 11`) to the `javac` call that compiles Piranha's module descriptors?

**Reply by Thihup** [reply]
Oh, well, at least we found the root cause.

**Reply by Thihup** [reply]
> Is it possible to add `--release 9` (or `--release 11`) to the `javac` call that compiles Piranha's module descriptors?

Probably, but it won't be the "truth" as the users still would need JDK 15 (sooner JDK 16).

**Reply by sormuras** [reply]
That's totally fine, to require 15+ -- the issue here is definitely with the Java version Sonatype is running the `modulescanner` program with. As you can see in the linked issue, Bach's modules require 16 already and don't show up in this database either...
